File: greenhouse_aiot/backend/models/alert.py
"""Alert model — maps to the `alerts` table defined in D1 schema.sql."""

from datetime import datetime

from models import db


class Alert(db.Model):
    """Automated alert raised when a sensor reading violates threshold rules."""

    __tablename__ = "alerts"

    alert_id        = db.Column(db.Integer,     primary_key=True)
    device_id       = db.Column(db.Integer,     db.ForeignKey("devices.device_id"),          nullable=False)
    reading_id      = db.Column(db.BigInteger,  db.ForeignKey("sensor_readings.reading_id"))
    assigned_to     = db.Column(db.Integer,     db.ForeignKey("users.user_id"))
    alert_type      = db.Column(db.String(30),  nullable=False)
    severity        = db.Column(db.String(10),  nullable=False, default="medium")
    message         = db.Column(db.String(500), nullable=False)
    measured_value  = db.Column(db.Numeric(10, 2))
    threshold_value = db.Column(db.Numeric(10, 2))
    status          = db.Column(db.String(20),  nullable=False, default="open")
    resolved_at     = db.Column(db.DateTime)
    created_at      = db.Column(db.DateTime,    default=datetime.utcnow)

    VALID_TYPES: frozenset[str] = frozenset({
        "temperature", "humidity", "ph", "light",
        "co2", "soil_moisture", "device_offline", "prediction",
    })
    VALID_SEVERITIES: frozenset[str] = frozenset({"low", "medium", "high", "critical"})
    VALID_STATUSES:   frozenset[str] = frozenset({"open", "acknowledged", "resolved", "dismissed"})

    def to_dict(self) -> dict:
        """Serialise to a JSON-safe dict.

        Field name aliases follow D3 frontend conventions:
          type      ← alert_type
          measured  ← measured_value
          threshold ← threshold_value
        """
        return {
            "alert_id":       self.alert_id,
            "device_id":      self.device_id,
            "reading_id":     self.reading_id,
            "assigned_to":    self.assigned_to,
            "alert_type":     self.alert_type,
            "type":           self.alert_type,
            "severity":       self.severity,
            "message":        self.message,
            "measured_value": float(self.measured_value)  if self.measured_value  is not None else None,
            "measured":       float(self.measured_value)  if self.measured_value  is not None else None,
            "threshold_value": float(self.threshold_value) if self.threshold_value is not None else None,
            "threshold":      float(self.threshold_value) if self.threshold_value is not None else None,
            "status":         self.status,
            "resolved_at":    self.resolved_at.isoformat() if self.resolved_at else None,
            "created_at":     self.created_at.isoformat() if self.created_at else None,
        }

    def __repr__(self) -> str:
        return f"<Alert {self.alert_id} {self.alert_type} {self.severity}>"

File: greenhouse_aiot/spring-backend/src/main/java/com/aiot/greenhouse/model/Alert.java
package com.aiot.greenhouse.model;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * Alerta automática generada cuando las lecturas de sensores salen de los rangos óptimos.
 */
@Entity
@Table(name = "alerts")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Alert {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "alert_id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "device_id", nullable = false)
    private Device device;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reading_id")
    private SensorReading reading;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to")
    private User assignedTo;

    @Enumerated(EnumType.STRING)
    @Column(name = "alert_type", nullable = false, length = 30)
    private AlertType alertType;

    @Enumerated(EnumType.STRING)
    @Column(name = "severity", nullable = false, length = 10)
    @Builder.Default
    private Severity severity = Severity.MEDIUM;

    @Column(name = "message", nullable = false, length = 500)
    private String message;

    @Column(name = "measured_value", precision = 10, scale = 2)
    private BigDecimal measuredValue;

    @Column(name = "threshold_value", precision = 10, scale = 2)
    private BigDecimal thresholdValue;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 15)
    @Builder.Default
    private AlertStatus status = AlertStatus.OPEN;

    @Column(name = "resolved_at")
    private LocalDateTime resolvedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    /** Tipo de magnitud que disparó la alerta. */
    public enum AlertType {
        TEMPERATURE, HUMIDITY, PH, LIGHT, CO2, SOIL_MOISTURE, DEVICE_OFFLINE, PREDICTION
    }

    /** Nivel de criticidad de la alerta. */
    public enum Severity {
        LOW, MEDIUM, HIGH, CRITICAL
    }

    /** Estado del ciclo de vida de la alerta. */
    public enum AlertStatus {
        OPEN, ACKNOWLEDGED, RESOLVED, DISMISSED
    }
}
